File: Ex-18/c++/Universidade.cpp
#include "Universidade.hpp"

Universidade::Universidade() : nome(""), departamento(nullptr)
{
}

Universidade::Universidade(const string &nome, const Departamento &departamento) : nome(nome), departamento(nullptr)
{
}

Universidade::~Universidade()
{
}

const string &Universidade::getNome() const
{
  return this->nome;
}

void Universidade::setNome(const string &nome)
{
  this->nome = nome;
}

Departamento *Universidade::getDepartamento() const
{
  return this->departamento;
}

void Universidade::setDepartamento(Departamento *departamento)
{
  this->departamento = departamento;
}

File: Ex-21/ex_21_java/Universidade.java
public class Universidade {
    private String nome;
    private static final int MAX_DEPARTAMENTOS = 50;
    private Departamento[] departamentos;
    private int numDepartamentosAtuais;

    public Universidade(String nome) {
        this.nome = nome;
        this.departamentos = new Departamento[MAX_DEPARTAMENTOS];
        this.numDepartamentosAtuais = 0;
    }

    public String getNome() {
        return this.nome;
    }

    public boolean adicionarDepartamento(Departamento depto) {
        if (this.numDepartamentosAtuais < MAX_DEPARTAMENTOS) {
            this.departamentos[this.numDepartamentosAtuais] = depto;
            this.numDepartamentosAtuais++;
            
            // A universidade, ao aceitar o departamento, estabelece o vínculo reverso
            depto.setUniversidade(this);
            return true;
        } else {
            System.out.println("Erro: Universidade '" + this.nome + "' já atingiu o limite de " + MAX_DEPARTAMENTOS + " departamentos.");
            return false;
        }
    }

    /**
     * Método para listar os departamentos afiliados a esta universidade.
     */
    public void listarDepartamentos() {
        System.out.println("Departamentos da " + this.nome + ":");
        if (numDepartamentosAtuais == 0) {
            System.out.println("- Nenhum departamento cadastrado.");
            return;
        }
        for (int i = 0; i < this.numDepartamentosAtuais; i++) {
            System.out.println("- " + this.departamentos[i].getNome());
        }
    }
}